Added year-picker, by modifying a copy of the date-picker

joakimgk opened the pull request:
New attempt, this time done on the "next" branch. Checked that there are no lint errors, but I have not made any tests (in year-calendar/year-calendar.service.spec.ts). Hope this is a useable starting point, at least, to assess my contribution. The Year Picker works on the Demo page.

vlio20 commented:
@joakimgk, can you please post a link to the demo page?

vlio20 commented:
please note that you are also have errors in the Travis build

joakimgk commented:
Well, I mean the demo in demo/demo/demo.component.html (as included in index.html when the project is run)

joakimgk commented:
Yes, I noticed that. Could you please point me in the right direction? Are they complaining about missing tests?

vlio20 commented:
could you please create a github page? it's easy - because it's part of the build of this component.
you just need to run:
switch to release branch
npm run build:demo
git add -A
git commit -m "adding demo"
git subtree push --prefix dist origin gh-pages

joakimgk commented:
Did it (from my "next" branch, which is where my modifications are). This creates a demo page on my GitHub?

vlio20 commented:
yes, I will take a look at the build as soon as I can

joakimgk commented:
Thanks! So, where do you find this demo page, then? :)

vlio20 commented:
After posting it, it will be in your repo github pages page

joakimgk commented:
Did you find the demo page?

vlio20 commented:
Nope

On Oct 20, 2017 4:39 PM, "Joakim" <notifications@github.com> wrote:

> Did you find the demo page?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/vlio20/angular-datepicker/pull/261#issuecomment-338209373>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEVQsbxRieA1KUPg2AauQv-jzzOAwseMks5suKJ9gaJpZM4QAlnd>
> .
>

joakimgk commented:
What does "switch to release branch" mean? Since all my modifications are on the "next" branch, I expect this is where I should create the demo. However, "npm run build:demo" yields no modifications, so "git add -A" and "git commit" does nothing.

vlio20 commented:
the problem that on the next branch the demo artefacts are ignored, switch to `release`  branch and merge to it the next branch. Then you should be able to run `build:demo` and then you will be able to add the demo pages as they are not ignored in this page.

joakimgk commented:
I have tried to switch to the `release` branch, and merge in my changes from `next`, but there is just no way I am able to do the `push`. I've tried to make sure branches are up-to-date, but I still get errors when doing `git subtree push --prefix dist origin gh-pages`:

> error: failed to push some refs to 'https://github.com/joakimgk/angular-datepicker.git'
> hint: Updates were rejected because a pushed branch tip is behind its remote
> hint: counterpart. Check out this branch and integrate the remote changes
> hint: (e.g. 'git pull ...') before pushing again.
> hint: See the 'Note about fast-forwards' in 'git push --help' for details.

vlio20 commented:
You don't need to push, just the subtree command

joakimgk commented:
Sorry, I don't understand. The error comes from the command `git subtree push --prefix dist origin gh-pages`, which involves a push, right? Anyway, while struggling to make a demo page, I've also been looking at the failed build. I fixed a typo in my `year-calendar.component.spec.ts`, and, after adding `"es6", "dom"` to the `tsconfig.spec.json` file, there are only two remaining failing tests -- but those are in files I have not modified: 

- `demo.component.spec.ts` invokes an empty constructor of `DemoComponent`, but `demo.component.ts` has no empty constructor. I fixed this (?) by simply passing in `null`:
```
describe('Component: Demo', () => {
  it('should create an instance', () => {
    const component = new DemoComponent(null);
    expect(component).toBeTruthy();
  });
});
```

- a similar test failed in `date-picker.directive.spec.ts` (line 20), in invoking the `DatePickerDirective` constructor. I had to add another `null`, to account for the fifth argument (even though it is decorated with `@Optional`).

**So, if `npm test -- -watch=false` yields SUCCESS (on my "next" branch), would you say it's OK for me to attempt another pull request?**

vlio20 commented:
Correct. From the error I see that you need to pull first.

Regarding tests. Please note that there are e2e tests as well. Also note that Master and Next are both passing the build.

joakimgk commented:
Did it finally go through? :-)

vlio20 commented:
Not yet. I hope I will have time today or tomorrow

vlio20 commented:
Is there a demo already?

joakimgk commented:
No, the demo thing I just had to give up. I would want to fix it, but I'd need some very specific instructions. My status is, I'm on the next branch, which is in sync with the same brach in your repo (right?). However, as I said before, when I switch to the release branch, and merge next into that, then `git add -A` and `commit` says no files are modified and branches are up to date. Yet, there are problems when I try the `git subtree` command.

vlio20 commented:
You may need to pull the release branch before.

joakimgk commented:
Well, hopefully you can use this pull request to determine whether it's possible to merge this contribution into your project :-)

vlio20 commented:
I will clone your project and checkout the demo locally.

vlio20 commented on src/app/date-picker/date-picker-directive-config.model.ts:
```diff
@@ -1,9 +1,10 @@
 import {ITimeSelectConfig} from '../time-select/time-select-config.model';
 
-export interface IDatePickerDirectiveConfig extends IDayCalendarConfig, IMonthCalendarConfig, ITimeSelectConfig {
+export interface IDatePickerDirectiveConfig extends IDayCalendarConfig, IMonthCalendarConfig, IYearCalendarConfig, ITimeSelectConfig {
```
too long, please break to lines

vlio20 commented on src/app/demo/demo/demo.component.html:
```diff
@@ -292,6 +297,28 @@ <h2 class="dp-page-header">Angular Date/Time/Month Picker</h2>
+                        #monthPicker="ngModel"
+                        [(ngModel)]="date"
+                        (ngModelChange)="log($event)"
+                        [mode]="'year'"
```
you can just `mode="year"`

vlio20 commented on src/app/year-calendar/year-calendar-config.ts:
```diff
@@ -0,0 +1,28 @@
+  yearBtnFormat?: string;
+  yearBtnFormatter?: (day: Moment) => string;
+  yearBtnCssClassCallback?: (day: Moment) => string;
+  multipleYearsNavigateBy?: number;
```
what will this do in a year picker?

joakimgk replied:
Again, what are you refering to here? All the four lines, or any one in particular?

vlio20 replied:
`multipleYearsNavigateBy` & `showMultipleYearsNavigation`

vlio20 commented on src/app/year-calendar/year-calendar-config.ts:
```diff
@@ -0,0 +1,28 @@
+  yearFormat?: string;
+  yearFormatter?: (month: Moment) => string;
+  format?: string;
+  isNavHeaderBtnClickable?: boolean;
```
what will this do?

joakimgk replied:
Right -- I'll omit this, the formatting parts. Year can only be year/YYYY, as I'm sure is your point.  The year calendar is molded from the month calendar, at first very closely to get it running. This detailed feedback from you is precisely what I was hoping for, to help me finish the code.

joakimgk replied:
s

joakimgk replied:
Looking at this now, and I'm not sure what, specifically, you're refering to. As far as I can see, the Year Calendar Config needs to specify a year format string, even if it can't be anything else than `YYYY`. However, it seems I can do without the `yearFormatter`. Can you please elaborate?

vlio20 commented on src/app/year-calendar/year-calendar-config.ts:
```diff
@@ -0,0 +1,28 @@
+
+export interface IYearCalendarConfig extends IConfig,
+                                              ICalendar {
+
```
redundant

joakimgk replied:
OK

joakimgk replied:
... however, this is identical to `month-calendar-config.ts`. I don't understand your comment, can I simply remove the `IYearCalendarConfig` interface?

vlio20 commented on src/app/year-calendar/year-calendar.component.html:
```diff
@@ -0,0 +1,31 @@
+      [showRightNav]="showRightNav"
+      [showRightSecondaryNav]="showSecondaryRightNav"
+      [isLabelClickable]="componentConfig.isNavHeaderBtnClickable"
+      [showGoToCurrent]="shouldShowCurrent()"
```
please don't use getter functions in template, use variables instead, this was implemented in 2.6.0 cross all components

joakimgk replied:
But this is the same way in `month-calendar/month-calendar.component.html`, isn't it?

vlio20 replied:
this might be something I missed, thanks for pointing out. Will fix on the month calendar.

vlio20 commented on src/app/year-calendar/year-calendar.component.ts:
```diff
@@ -0,0 +1,254 @@
+  @HostBinding('class') @Input() theme: string;
+
+  @Output() onSelect: EventEmitter<IYear> = new EventEmitter();
+  @Output() onNavHeaderBtnClick: EventEmitter<null> = new EventEmitter();
```
should be void?

joakimgk replied:
Please explain

vlio20 replied:
which one more relevant `EventEmitter<null>`  or EventEmitter<void>? are you emitting null or undefined?

vlio20 commented on src/app/year-calendar/year-calendar.component.ts:
```diff
@@ -0,0 +1,254 @@
+
+  onLeftNav() {
+    this.currentDateView = this.currentDateView.clone().subtract(12, 'year');
+    this.yearMatrix = this.yearCalendarService.generateYear(this.componentConfig, this.currentDateView, this.selected);
```
should be changed when currentDateView is changed, please see other components

joakimgk replied:
Isn't this equivalent to the same method, in `month-calendar.component.ts`? I don't understand your comment.

vlio20 replied:
take a look at https://github.com/vlio20/angular-datepicker/blob/master/src/app/month-calendar/month-calendar.component.ts#L86

joakimgk replied:
Yes, and my `currentDateView` method in `year-calendar.component.ts` does the same (please note I renamed the `this.yearMonths` to `this.yearMatrix`, for consistency): 
<snip>
    set currentDateView(current: Moment) {
        this._currentDateView = current.clone();
        this.yearMatrix = this.yearCalendarService
          .generateYear(this.componentConfig, this._currentDateView, this.selected);
</snip>

joakimgk replied:
(didn't find a multi-line code formatting)

vlio20 commented on src/app/year-calendar/year-calendar.component.ts:
```diff
@@ -0,0 +1,254 @@
+
+  onRightNav() {
+    this.currentDateView = this.currentDateView.clone().add(12, 'year');
+    this.yearMatrix = this.yearCalendarService.generateYear(this.componentConfig, this.currentDateView, this.selected);  // remove?
```
remove???

joakimgk replied:
As above -- I've created the year calendar from the design of the month calendar. In `month-calendar.component.ts`, there is a, perhaps, strange difference between methods `onLeftNav()` and `onRightNav()`; only the former actually updates the `yearMonths` object. Why doesn't the "right" handler also update the object? I opted to update `yearMatrix` in both methods, but marked the code to draw your attention.

joakimgk replied:
Any thoughts on this? :)

vlio20 replied:
this is very odd... I need to check. Sorry for the long responses...

vlio20 commented on src/app/year-calendar/year-calendar.component.ts:
```diff
@@ -0,0 +1,254 @@
+    if (isOutsideRange) {
+      navigateBy = this.componentConfig.max.year() - this.currentDateView.year();
+    }
+    this.currentDateView = this.currentDateView.clone().add(navigateBy, 'year');
```
add empty line before this row

joakimgk replied:
OK

vlio20 commented on src/app/year-calendar/year-calendar.component.ts:
```diff
@@ -0,0 +1,254 @@
+      navigateBy = this.componentConfig.max.year() - this.currentDateView.year();
+    }
+    this.currentDateView = this.currentDateView.clone().add(navigateBy, 'year');
+    this.yearMatrix = this.yearCalendarService.generateYear(this.componentConfig, this.currentDateView, this.selected);  // remove?
```
remove??

joakimgk replied:
-- see above --

vlio20 commented on src/app/year-calendar/year-calendar.service.spec.ts:
```diff
@@ -0,0 +1,25 @@
+      providers: [YearCalendarService, UtilsService]
+    });
+  });
+
```
many redundant empty lines

joakimgk replied:
OK

vlio20 commented on src/app/year-calendar/year-calendar.service.ts:
```diff
@@ -0,0 +1,97 @@
+        };
+
+        index.add(1, 'year');
+        return _year;
```
add line break before return

joakimgk replied:
OK

vlio20 commented on src/app/year-calendar/year-calendar.service.ts:
```diff
@@ -0,0 +1,97 @@
+
+  getHeaderLabel(config: IYearCalendarConfig, year: Moment): string {
+    return null;
+    /*
```
please remove if not required

joakimgk replied:
OK. You agree this is not needed? I considered making a header label of sorts, "2017-2028", but I'm not sure that's better than simply no header.

vlio20 commented on src/app/year-calendar/year-calendar.service.ts:
```diff
@@ -0,0 +1,97 @@
+    if (config.yearBtnFormatter) {
+      return config.yearBtnFormatter(year);
+    }
+    return year.format(config.yearBtnFormat);
```
new line before return

joakimgk replied:
OK

joakimgk commented:
Thanks for the very helpful feedback! Not sure why the build failed this time, to be honest, but I'll fix the things you point out and make a new push tomorrow. Hopefully it will build, then ;)

vlio20 commented:
Thanks for the PR. 
There are some parts that should be added:
1. e2e
2. documentation
3. Integration with month picker (same as day and month). There should be a way to jump from day to month to day (all should be available via configuration).

vlio20 commented on src/app/year-calendar/year-calendar-config.ts:
```diff
@@ -6,7 +6,6 @@ export interface IConfig {
   isYearDisabledCallback?: (date: Moment) => boolean;
   allowMultiSelect?: boolean;
   yearFormat?: string;
-  yearFormatter?: (month: Moment) => string;
```
why did you remove that?

joakimgk replied:
Ah, I think I need more specific instructions ;-) Hope you can point me in the right direction, especially wrt hooking the year calendar up to the month and date calendars. If the other things you point out are minor things, please help me get started :-)

joakimgk commented:
Is there perhaps any technical documentation on how the datepicker is set up? I mean, how the different calendars work by themselves, and how they work together? I haven't had time to pick apart the existing code, and only have a working understanding (of the parts encountered while making the year calendar).

vlio20 commented:
There is a documentation for each component, but you will understand each configuration by its template

vlio20 commented on .gitignore:
```diff
@@ -36,3 +36,5 @@ testem.log
 
 dist
+
+*.bak
```
?

vlio20 commented on src/app/date-picker/date-picker-config.model.ts:
```diff
@@ -20,12 +21,14 @@ export interface IConfig {
                                            IMonthCalendarConfig,
+                                           IYearCalendarConfig,
                                            ITimeSelectConfig {
 
```
redundant
